### PR TITLE
feature/eng-4292-add-shift-assignment-and-shift-creation-as-templates-in

### PR DESCRIPTION
**Pull request by nilsbeck**
# Description

Adds templates for `shift-planning` and `shift-assignment` to sdk.

## Changes

new folders and template files in

* `templates/shift-planning`
* `templates/shift-assignment`

### CODE REVIEW

File: ruff.toml [hunk 1]
```diff
@@ -14,3 +14,7 @@ select = [
 # Rule configuration.
 line-length = 120
 
+[extend-per-file-ignores]
+"templates/shift-assignment-ortools/main.py" = ["C901"]
+"templates/shift-planning-ortools/main.py" = ["C901"]
+"templates/shift-assignment-pyomo/main.py" = ["C901"]
```
Comment (merschformann) [inline ruff.toml:20]: I should have started using ruff for that tech talk as well. Are these easy fixes? :thinking: 
Probably nothing we want to spend time on though.

File: templates/knapsack-gosdk/README.md
Comment (merschformann) [inline templates/knapsack-gosdk/README.md:1]: Big fan of this restructure. We have too many overlapping apps (and apparently will add further ones). I think a naming rule like "name-type" is a good start. :blush:

File: templates/order-fulfillment-gosdk/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Nextmv MIP knapsack template
```
Comment (nmisek) [inline templates/order-fulfillment-gosdk/README.md]:
```suggestion
# Nextmv order fulfillment template
```

File: templates/order-fulfillment-gosdk/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Run the command below to see if everything works as expected:
+
+```bash
+nextmv sdk run . -- -runner.input.path input.json \
+  -runner.output.path output.json -solve.duration 10s
+```
+
+A file `output.json` should have been created with the optimal knapsack
+solution.
```
Comment (nmisek) [inline templates/order-fulfillment-gosdk/README.md]:
```suggestion
A file `output.json` should have been created with a solution to the order fulfillment problem.
```

File: templates/routing-ortools/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Routing with OR-Tools
+
+This app solves a routing problem using [OR-Tools][or-tools]. Given a distance
+matrix (and with it indirectly the stops), a number of vehicles and a depot.
```
Comment (nmisek) [inline templates/routing-ortools/README.md]:
```suggestion
This app solves an unconstrained routing problem using [OR-Tools][or-tools]. Given a distance
matrix (and with it indirectly the stops), a number of vehicles and a depot.
```

File: templates/routing-ortools/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+1. Run the command below to check that everything works as expected:
+
+    ```bash
+    python3 main.py -input input.json -output output.json -duration 30
+    ```
+
+1. A file `output.json` should have been created with the optimal knapsack
+   solution.
+
```
Comment (nmisek) [inline templates/routing-ortools/README.md]:
```suggestion
1. A file `output.json` should have been created with a solution to the routing problem. 

```

File: templates/shift-assignment-ortools/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    ```
+
+1. Run the command below to check that everything works as expected:
+
+    ```bash
+    python3 main.py -input input.json -output output.json -duration 30
+    ```
+
+1. A file `output.json` should have been created with the optimal knapsack
+   solution.
```
Comment (nmisek) [inline templates/shift-assignment-ortools/README.md]:
```suggestion
1. A file `output.json` should have been created with a solution to the shift assignment problem.
```

File: templates/shift-assignment-pyomo/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Run the command below to check that everything works as expected:
+
+    ```bash
+    python3 main.py -input input.json -output output.json -duration 30
+    ```
+
+1. A file `output.json` should have been created with the optimal knapsack
+   solution.
+
```
Comment (nmisek) [inline templates/shift-assignment-pyomo/README.md]:
```suggestion
1. A file `output.json` should have been created with a solution to the shift assignment problem.

```

File: templates/shift-planning-ortools/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Run the command below to check that everything works as expected:
+
+    ```bash
+    python3 main.py -input input.json -output output.json -duration 30
+    ```
+
+1. A file `output.json` should have been created with the optimal knapsack
+   solution.
+
```
Comment (nmisek) [inline templates/shift-planning-ortools/README.md]:
```suggestion
1. A file `output.json` should have been created with a solution to the shift planning problem.

```

**Review by nmisek — DISMISSED** [review]
Requested some small changes to READMEs

File: templates/routing-ortools/README.md (hunk 5 above)
Comment (merschformann) [reply templates/routing-ortools/README.md]: image

File: templates/routing-ortools/main.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            )
+        stops.append(manager.IndexToNode(index))
+        route = {
+            "vehicle": vehicle_id,
+            "distance": route_distance,
+            "stops": stops,
+        }
+        routes.append(route)
+        max_route_distance = max(route_distance, max_route_distance)
+        activated_vehicles += 1
```
Comment (merschformann) [inline templates/routing-ortools/main.py]: This means that all vehicles are always activated? :thinking:

File: templates/routing-ortools/main.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import sys
+import time
+from typing import Any
+
+from ortools.constraint_solver import pywrapcp, routing_enums_pb2
+
+
+def main():
+    """Entry point for the template."""
+    parser = argparse.ArgumentParser(description="Solve problems with OR-Tools.")
```
Comment (merschformann) [inline templates/routing-ortools/main.py]:
```suggestion
    parser = argparse.ArgumentParser(description="Solves VRPs with a max distance constraint.")
```

File: templates/routing-ortools/main.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    distance_dimension.SetGlobalSpanCostCoefficient(100)
+
+    # Setting first solution heuristic.
+    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
+    search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
+    search_parameters.time_limit.FromSeconds(duration)
+
+    # Solve the problem.
+    start_time = time.time()
+    solution = routing.SolveWithParameters(search_parameters)
```
Comment (merschformann) [inline templates/routing-ortools/main.py:78]: We need to handle the case where no solution is found. E.g., the solver ran out of time for a highly constrained input or the input is infeasible. In both cases the returned solution will simply be `None`, I believe.

File: templates/routing-ortools/main.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    # Define cost of each arc.
+    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)
+
+    # Add Distance constraint.
+    dimension_name = "Distance"
+    routing.AddDimension(
+        transit_callback_index,
+        0,  # no slack
+        3000,  # vehicle maximum travel distance
```
Comment (merschformann) [inline templates/routing-ortools/main.py]: We should make this part of the input (e.g.: `"max_distance: 3000"`).

File: templates/shift-assignment-ortools/README.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
+# Shift assignment with OR-Tools
+
+This app solves a shift asssignment problem using [OR-Tools][or-tools]. Given a
+set of previously planned shifts, in this app we assign workers to those shifts,
```
Comment (merschformann) [inline templates/shift-assignment-ortools/README.md:4]: We could link to the shift-planning app here.

File: templates/shift-assignment-pyomo/main.py [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+        "-duration",
+        default=30,
+        help="Max runtime duration (in seconds). Default is 30.",
+        type=int,
+    )
+    parser.add_argument(
+        "-provider",
+        default="cbc",
+        help="Solver provider. Default is cbc.",
+    )
```
Comment (merschformann) [inline templates/shift-assignment-pyomo/main.py:50]: This is interesting. We should probably do this for ortools as well, if we do it for pyomo, right?

File: templates/shift-planning-pyomo/main.py [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+    obj_expr += model.shift_cost
+    model.obj = Objective(expr=obj_expr, sense=minimize)
+
+    # Constraints
+
+    # We need to make sure that all demands are covered (or track under supply).
+    for p in periods:
+        constraint_name = f"DemandCover_{p.start_time}_{p.end_time}_{p.qualification}"
+        if constraint_name in model.component_map(Constraint):
+            model.del_component(constraint_name)
```
Comment (merschformann) [inline templates/shift-planning-pyomo/main.py]: Why is this needed? :thinking:

File: ruff.toml (hunk 1 above)
Comment (nilsbeck) [reply ruff.toml:20]: I'd leave it for not. I fixes a couple of other things

File: templates/routing-ortools/main.py (hunk 10 above)
Comment (nilsbeck) [reply templates/routing-ortools/main.py]: would not do that. we want to point out that it is OR-Tools. not that there is a distance constraint

File: templates/shift-planning-pyomo/main.py (hunk 15 above)
Comment (nilsbeck) [reply templates/shift-planning-pyomo/main.py]: yeah, not sure tbh. There was an error of doubled constraints and a quick search gave this solution. no idea if this is normal or not...

**Comment by nilsbeck** [comment]
> Requested some small changes to READMEs

yeah, found some more knapsacks and fixed them

File: templates/routing-ortools/main.py (hunk 10 above)
Comment (merschformann) [reply templates/routing-ortools/main.py]: Ok, fair. The distance constraint is not so important, but "solve problems" sounds a little too generic, right?

Comment (nilsbeck) [reply templates/routing-ortools/main.py]: ok, updated

**Review by merschformann — APPROVED** [review]
